Function composition and negation, take 2

andyferris opened the pull request:
Something strange happened to my previous PR #17119, and github closed it. 

This is the basic version with just an anonymous function version of `∘` and a definition of function negation. This is quite related to #17155.

@StefanKarpinski feel free to just take docstrings and tests (in general - is it better to make a PR to your PR?)

StefanKarpinski committed:
function negation and composition: !f, f∘g

Commit:
Added docstrings and tests for compositions

tkelman commented on base/operators.jl:
```diff
@@ -328,9 +328,31 @@ eltype(::Type{Any}) = Any
 
+"""
+    ∘(f, g)
+    f ∘ g
```
would need to put the signatures into the RST docs so `make docs` populates the docstrings there, see details in contributing.md

andyferris replied:
I added `.. function:: ∘(f, g)` to base.rst and, although it worked for `.. function:: !(f::Function)`, I get the following:
```
WARNING: missing docs for signature:

    ∘(f, g)
```
Is this something to do with unicode operators?

andyferris replied:
@tkelman (not sure if you'd get auto-pinged or not, here)

tkelman replied:
Yes, no need to explicitly ping me if I've already commented on this issue. It may be a unicode issue where you might have to explicitly declare unicode operators somewhere for RST/Sphinx (I'm not sure where that is though), or it may have to do with the separate signatures being on multiple lines.

andyferris replied:
OK, yes it is the latter

andyferris replied:
That's a little unfortunate though. Infix operators should allow an infix form. Even typing `?<:` in the REPL gives help for `<:(T1, T2)` - have you ever seen a single line of Julia code written like that??

nalimilan replied:
At least it shows you that operators are just standard functions (OK, not true for `<:`) and that you can add a method using the standard syntax. But we should definitely improve these docstrings by adding the infix forms on the second line at the minimum.

MichaelHatherly replied:
@andyferris did you include the entire signature, i.e.

```rst
.. function:: ∘(f, g)
              f ∘ g

```

or just the first line `.. function:: ∘(f, g)`? Signature matching requires the entire first code block, not just the first line of it.

TotalVerb commented:
How is performance? Does the splatting penalty apply here?

andyferris commented:
@TotalVerb Good question! Simple check says "no". @StefanKarpinski do you know where we are up to here? I have seen some issues that discuss this and I thought there was a simple fix in place for `f(x...) = g(x...)` patterns (although this is epsilon more complicated than that).

andyferris commented:
Indeed, if we can't work around it, then maybe we have to follow @TotalVerb's suggestion until the splatting is fixed. I would suggest generating versions up to `MAX_TUPLE_SPLAT` (currently 16), for consistency.

andyferris commented:
> however it probably slows down ∘ itself.

A functor approach might help with that... define 17 methods globally on calling a `ComposedFunction{F,G}` and not generate them every time someone calls `∘`.

Commit:
rst for docs

TotalVerb commented:
I don't like calling these "functors", especially since (from the user's perspective) they're not meaningfully different from plain old closures.

I guess you mean something like this? I did not get a chance to look at your last PR, but I assume it was a similar approach?

```
immutable ComposedFunction{F,G}
    f::F
    g::G
end
```

TotalVerb commented:
Actually, since `h` is a closure, the methods should already be added only once—not once on each call—so the functor should be unnecessary.

andyferris commented:
OK, yes, functor is a strange name. I meant exactly the thing your wrote, `ComposedFunction`.

I agree the `h` as a closure will work fast and "only be constructed once per composed function". I was suggesting that we only define the 17 splatting methods "once per session of Julia".

I actually don't care if there is a once-off overhead to `∘`. Just one more reason to fix the splatting penalty!

TotalVerb commented:
@andyferris What I meant was that my assessment was incorrect, and there is no such cost when creating a composed function, because methods are on the (abstract) closure type. So the performance of a "functor" is equivalent to the performance of a closure.

Commit:
Changed to closure approach: no splatting penalty

Decided to make up to 16 inputs fast, to match MAX_TUPLE_SPLAT.
Conveniently, closures also allow for introspection of f and g.

andyferris commented:
Hmm... its a bit uglier, but we now have no splatting penalty!

I just discovered closures let us have access to the two things we are composing, allowing for introspection. Cool!

(Closures also seem to be a type of "generated type", in the sense the field names and number depend on how you define them. Super cool!)

nalimilan commented on base/operators.jl:
```diff
@@ -339,18 +339,80 @@ Creates a composition of two functions (or functor objects), such that
+equivalent to `(x...) -> !f(x...)`.
 """
-!(f::Function) = (x...)->!f(x...)
+function !(f::Function)
```
Can't you just define this as `! ∘ f` to avoid repeating the same machinery as above?

andyferris replied:
I would like to, and I tried that first, but I got some kind of recursion when I tried to compile Julia. I didn't have time to investigate the cause of that yet, so I just used copy-paste.

TotalVerb replied:
I'm pretty sure that's because `! ∘ f` doesn't parse correctly. We probably need `(!) ∘ f`.

andyferris replied:
That worked...

Commit:
Fixed typo in test

Commit:
Simplified definition of function negation

andyferris commented:
> I'm pretty sure that's because ! ∘ f doesn't parse correctly. We probably need (!) ∘ f. This is a little ugly, which makes me like !f syntax a little more.

We will probably see this with other operators. This is an argument for even more overloads of operators on functions, as in #17155... it would even suggest doing it for *every* operator. Not sure if I like that or not though!

mweastwood commented on test/operators.jl:
```diff
@@ -55,3 +55,6 @@ let xs = [[i:i+4;] for i in 1:10]
     end
 end
+
+@test ((x -> x+1) ∘ (x _> 2x))(5) == 11
```
typo `@test ((x -> x+1) ∘ (x -> 2x))(5) == 11`

andyferris replied:
Thanks

andyferris commented:
@StefanKarpinski @JeffBezanson Where are we on this, and what is the probability of 0.5 inclusion? 

Between here and #17155, the `∘` seems pretty well received. There are some contentions with unary operators like `!` and much, much stronger contentions with binary operators like `&`, on the basis of not knowing where to draw the line on auto-composition.

We could maybe try `!` as an experiment in 0.5? Or drop it completely?

tkelman commented on base/operators.jl:
```diff
@@ -328,9 +328,61 @@ eltype(::Type{Any}) = Any
+"""
+    ∘(f, g)
+
+Creates a composition of two functions (or functor objects), such that
```
this pr doesn't include functor objects any more, does it?

andyferris replied:
No, it uses closures. But ∘ *accepts* functors (any callable) as arguments.

andyferris replied:
I'll edit it to make it clear.

andyferris committed:
Clarify docstring

oxinabox commented:
I noted today a nice use case for this: Eager-ifying lazy operations


```
function lazyload_semcor(tagdir_path::AbstractString)
    Task(...)
end

load_semcor(tagdir_path::AbstractString) = collect(lazyload_semcor(tagdir_path))
```

This would become:

```
load_semcor = collect∘lazyload_semcor
```

Which is nice.

andyferris commented:
Yes, that's right. I quite enjoy using these already in our *CoordinateTransformations* for similar reasons. You gain a lot of control on how things are done with little user effort.

The reason my original PR had a `ComposedFunction` object was to allow for even more in-depth introspection where composing certain things together might do something non-trivial - an example of the top of my head is `map ∘ reduce -> mapreduce` (because they take multiple arguments that doesn't work as I wrote it, obviously, but you can see we can skip an allocation and go straight to iteration - in other words going the opposite way you said, from eager to lazy, but only where appropriate).

tkelman commented:
closing in favor of #17155
